### PR TITLE
Check for deferred reports after reporting current state

### PR DESCRIPTION
**Pull request by 20k-ultra**
This ensures any reports attempted to be sent when a report is already being sent will be checked once the first (pending/active) report is completed.

This is so that any changes that occur during a report make it to the cloud in a more timely manner.

### CODE REVIEW

File: src/api-binder/report.ts [hunk 1]
```diff
@@ -139,11 +139,26 @@ export async function startReporting() {
...
+				// Check if a report was requested while this one was processing
+				if (reportDeferred) {
+					reportDeferred = false;
+					// Try to report again in case something changed while the last report was running
+					doReport();
+				}
+			}
+		} else {
+			// Defer this report so we can try again once the current one is completed
+			reportDeferred = true;
```
Comment (thgreasi) [inline src/api-binder/report.ts:161]: You could also clear this inside `getStateAndReport()` , right below the `getCurrentForReport()`
See: https://github.com/balena-os/balena-supervisor/blob/e58683537afcc716d785c7632cb3ed3e1a22fdfa/src/api-binder/report.ts#L74

This way the deferred report is discarded, if the `getCurrentForReport()` is called after the change event, which means that the new data ended up being send already (eg b/c of a backoff).

File: src/api-binder/report.ts [hunk 2]
```diff
@@ -139,11 +139,26 @@ export async function startReporting() {
...
-			await reportCurrentState(reportConfigs);
-			reportPending = false;
+			try {
+				// Report changes in the current state
+				await reportCurrentState(reportConfigs);
+			} finally {
+				reportPending = false;
+				// Check if a report was requested while this one was processing
+				if (reportDeferred) {
+					reportDeferred = false;
```
Comment (thgreasi) [inline src/api-binder/report.ts:154]: I like to think about this variable as `shouldReportNewerState` and we move this assignment right after the `getCurrentForReport()`, we should be fine.

Comment (20k-ultra) [reply src/api-binder/report.ts]: went with `checkForNewerState` but not 100% on it yet.

**Comment by 20k-ultra** [comment]
might use https://github.com/balena-os/balena-supervisor/pull/1962 instead

**Comment by thgreasi** [comment]
@20k-ultra I have a feeling that this is more accurate thatn #1962, since a throttle can never know the internal state/progress of the exponential backoff mechanism, which we really need to know. I think this satisfies the requirements of the synchronization scheme that we try to achieve more accurately, regardless of not using a re-usable & generic `throttle` function.
It also works exactly how we describe the solution, which imo is a good sign.

**Pull request closed without merging**